src/process/message_processor.go

```diff
@@ -1,48 +1,101 @@
 package process
 
 import (
+	"BGPAlert/analyze"
 	"BGPAlert/common"
-	"errors"
+	"BGPAlert/config"
 	"fmt"
+	"strconv"
 	"time"
 )
 
-// Constantly reads channel of messages and stores them in Window objects to send through windowChannel for analysis
-func ProcessBGPMessages(msgChannel chan common.BGPMessage, windowChannel chan common.Window) error {
-	var bucketMap = make(map[time.Time][]common.BGPMessage)
+// Constantly read messages from channel, build up windows with frequency maps, calling analysis on windows when full
+func ProcessBGPMessages(msgChannel chan common.BGPMessage, config *config.Configuration) error {
+	// Parse windowSize from config
+	var maximumBuckets int
+	// If windowSize can't be parsed to an int -> just default to 30
+	if windowSize, err := strconv.ParseInt(config.WindowSize, 10, 64); err != nil {
+		fmt.Println("Inputted config size is not a number, defaulting to 30")
+		maximumBuckets = 30
+	} else {
+		maximumBuckets = int(windowSize)
+	}
+
+	// Each bucket is 60s
+	parseDuration := 60 * time.Second
+	maximumTimespan := time.Duration(maximumBuckets*60) * time.Second
+
+	window := common.Window{
+		Filter:    "none",
+		BucketMap: make(map[time.Time][]common.BGPMessage),
+	}
 
+	// Read messages from channel
 	for msg := range msgChannel {
-		windowSize, err := time.ParseDuration("60s")
-		if err != nil {
-			return errors.New("error creating parsing window, " + err.Error())
-		}
+		// Round timestamp to size of parseDuration to place in correct bucket
+		messageBucket := msg.Timestamp.Truncate(parseDuration)
+
+		// If a bucket with this timestamp doesn't exist, need to create one
+		if _, ok := window.BucketMap[messageBucket]; !ok {
+			fmt.Println("----------------------------------------------------------------------------------")
+
+			// If there's at least one timestamp, may have to fill in missing zeroes in map
+			if len(window.BucketMap) > 1 {
+				// Get minimum timestamp from map
+				minTimestamp := getBucketMapMin(window.BucketMap)
+
+				// Walk through map and see if any zeroes are missing
+				for tempTimestamp := minTimestamp; tempTimestamp.Before(messageBucket); tempTimestamp = tempTimestamp.Add(time.Minute) {
+					if _, ok := window.BucketMap[tempTimestamp]; !ok {
+						fmt.Println("Appended a 0 to map at ", tempTimestamp)
+						window.BucketMap[tempTimestamp] = make([]common.BGPMessage, 0)
+					}
+				}
+			}
+
+			// If we at least maximumBuckets in bucketMap we can perform analysis
+			if len(window.BucketMap) >= maximumBuckets {
+				fmt.Println("len(window.BucketMap):", len(window.BucketMap), " maximumBuckets:", maximumBuckets)
```

Comment:
nit pick - you can use fmt.Sprintf() [https://pkg.go.dev/fmt#Sprintf](https://pkg.go.dev/fmt#Sprintf)